src/web-extensions.c

```diff
@@ -0,0 +1,840 @@
+                gint status,
+                gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  web_extensions_session->child_pid = -1;
+  web_extensions_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  /* This regexp comes from the Mozilla documentation on valid native
+     messaging host names:
+
+     https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#native_messaging_manifests
+
+     That is, one or more dot-separated groups composed of
+     alphanumeric characters and underscores.
+  */
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (WebExtensionsSessionMode mode)
+{
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  search_path = g_ptr_array_new_with_free_func (g_free);
+  switch (mode)
+    {
+    case WEB_EXTENSIONS_SESSION_MODE_CHROMIUM:
+      /* Chrome and Chromium search paths documented here:
+       * https://developer.chrome.com/docs/extensions/nativeMessaging/#native-messaging-host-location
+       */
+      /* Add per-user directories */
+      g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "google-chrome", "NativeMessagingHosts", NULL));
+      g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "chromium", "NativeMessagingHosts", NULL));
+      /* Add system wide directories */
+      g_ptr_array_add (search_path, g_strdup ("/etc/opt/chrome/native-messaging-hosts"));
+      g_ptr_array_add (search_path, g_strdup ("/etc/chromium/native-messaging-hosts"));
+      /* And the same for xdg-desktop-portal's configured prefix */
+      g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "/opt/chrome/native-messaging-hosts"));
+      g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "/chromium/native-messaging-hosts"));
+      break;
+
+    case WEB_EXTENSIONS_SESSION_MODE_MOZILLA:
+      /* Firefox search paths documented here:
+       * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#manifest_location
+       */
+      /* Add per-user directories */
+      g_ptr_array_add (search_path, g_build_filename (g_get_home_dir (), ".mozilla", "native-messaging-hosts", NULL));
```

Comment:
We might want to also handle `$XDG_CONFIG_HOME/mozilla` here?

Reply:
@lissyx
Your comment is related to https://bugzilla.mozilla.org/show_bug.cgi?id=1903246 right?

Is there canonical documentation that documents this expected location?

Reply:
yes, it will be taken care of https://bugzilla.mozilla.org/show_bug.cgi?id=1905711, but i dont think we can update MDN so soon yet without any expected timeline, and i wanted to make sure there was a note of that somewhere

Reply:
Even if XDG_CONFIG_HOME were to be supported, then developers of native messaging components should still continue to write in the "old" location to support environments that do not support XDG_CONFIG_HOME. Therefore I think that it is acceptable to proceed with the current logic as is, and handle XDG_CONFIG_HOME later.

Do you agree with that assessment?

Reply:
Obviously, both will have to work, that's what we are already doing on Firefox side, only new profiles would default to `$XDG_CONFIG_HOME` use

Reply:
> Obviously, both will have to work, that's what we are already doing on Firefox side, only new profiles would default to `$XDG_CONFIG_HOME` use. I did mention « also » at first, i was not planning on replacing the path just having to dig into the code and see how to handle both

I was referring to external application developers, that they can't only support XDG_CONFIG_HOME, and that therefore it may be acceptable for the portal to not recognize XDG_CONFIG_HOME.

The timeline of shipping support for XDG_CONFIG_HOME is unclear to me. If it is "relatively soon", say a few months at most, then I suppose that we can integrate it in the portal right away.

I asked for documentation or a description of the XDG_CONFIG_HOME feature because that would make it more obvious what the desired order of lookup should be.

Reply:
> > Obviously, both will have to work, that's what we are already doing on Firefox side, only new profiles would default to `$XDG_CONFIG_HOME` use. I did mention « also » at first, i was not planning on replacing the path just having to dig into the code and see how to handle both
> 
> I was referring to external application developers, that they can't only support XDG_CONFIG_HOME, and that therefore it may be acceptable for the portal to not recognize XDG_CONFIG_HOME.

yes, that's why we try legacy first

> 
> The timeline of shipping support for XDG_CONFIG_HOME is unclear to me. If it is "relatively soon", say a few months at most, then I suppose that we can integrate it in the portal right away.

there's no hardcoded deadline yet, a few months might depend on the definition of "few" :), but it's obviously not going to land within the next weeks, and we probably dont want to wait years. So we definitively dont want to wait too much to integrate in the portal to make sure it has time to reach end users. Might also have to prepare backports patches for distros in case that is needed?

> 
> I asked for documentation or a description of the XDG_CONFIG_HOME feature because that would make it more obvious what the desired order of lookup should be.

canonical point of entry is still the (rather long) bug https://bugzilla.mozilla.org/show_bug.cgi?id=259356. But to summarize quickly:
 - `MOZ_LEGACY_HOME=1`: we continue with the existing behavior
 - `~/.mozilla/` exists?: we continue with the existing behavior
 - otherwhise we `$XDG_CONFIG_HOME`

Reply:
> we probably dont want to wait years

Please don't. Flatpak browsers have been crippled for far too long.

Reply:
> > we probably dont want to wait years
> 
> Please don't. Flatpak browsers have been crippled for far too long.

before this gets out of control, there's 0 reason to hold the current PR for `$XDG_CONFIG_HOME` handling.

Reply:
To recap, the XDG_CONFIG_HOME handling effort in Firefox (https://bugzilla.mozilla.org/show_bug.cgi?id=259356) is about making sure that Firefox adheres to the platform conventions as specified in XDG's Basedir spec.

In the context of Firefox only, this is about the ability to read and write to an application-specific location, located at `$XDG_CONFIG_HOME/mozilla/` instead of `~/.mozilla/`. For compatibility reasons, the lookup order is `~/.mozilla/` first, then `$XDG_CONFIG_HOME/mozilla/`. Firefox may create files if they do not exist yet, and to force the current default of creating in `~/.mozilla/`, `MOZ_LEGACY_HOME=1` can be set (if unset, `$XDG_CONFIG_HOME/mozilla/` would be created).

In the context of nativeMessaging and Firefox, it is relevant to know that this location is written to by external applications. Firefox treats them as read-only. Or, in the context of this portal, the portal reads from here and notifies Firefox of the discovered external native messaging applications.

The filesystem layout as seen by Firefox (inside the sandbox of the Snap package) can be independent of the filesystem layout outside of the sandbox (as seen by the portal). Therefore the mechanism to look up native messaging paths can be independent too.

Since the native messaging portal consumes the locations in a read-only fashion, there is no compatibility risk with unconditionally including `$XDG_CONFIG_HOME/mozilla/native-messaging-hosts` in the search_path (at the end). The only potential change in behavior is that the portal would be able to find and offer a native messaging host application, where Firefox would otherwise not be able to (until the patch stack associated with https://bugzilla.mozilla.org/show_bug.cgi?id=1905711 lands).

Reply:
Ack, added a commit to also include `$XDG_CONFIG_HOME/mozilla/native-messaging-hosts`, after `~/.mozilla/native-messaging-hosts`.